lib/plausible/billing/quota.ex

```diff
@@ -162,10 +177,30 @@ defmodule Plausible.Billing.Quota do
     ]
 
     Enum.reduce(queries, [], fn {feature, query}, acc ->
-      if Plausible.Repo.exists?(query), do: [feature | acc], else: acc
+      if Plausible.Repo.exists?(query), do: acc ++ [feature], else: acc
     end)
   end
 
+  def ensure_can_subscribe_to_plan(user, %Plan{} = plan) do
+    case exceeded_limits(usage(user), plan) do
+      [] -> :ok
+      exceeded_limits -> {:error, %{exceeded_limits: exceeded_limits}}
+    end
+  end
+
+  def ensure_can_subscribe_to_plan(_user, nil), do: :ok
+
+  def exceeded_limits(usage, %Plan{} = plan) do
+    for {usage_field, limit_field} <- [
+          {:monthly_pageviews, :monthly_pageview_limit},
+          {:team_members, :team_member_limit},
+          {:sites, :site_limit}
+        ],
+        !within_limit?(Map.get(usage, usage_field), Map.get(plan, limit_field)) do
```

Comment:
```suggestion
        !within_limit?(Map.fetch!(usage, usage_field), Map.fetch!(plan, limit_field)) do
```

just in case, to not let it fail silently and cause issues down the line perhaps?

Reply:
Sorry, just merged a sec before seeing your comment. It's a fair point. I can make a note and change in the next PR perhaps